java/impl/src/main/java/net/sf/mmm/code/impl/java/source/maven/JavaSourceProviderUsingMaven.java

```diff
@@ -30,170 +31,385 @@
+            String eclipseApiLocation = "";
+            for (String segment : segments) {
+                eclipseApiLocation = eclipseApiLocation + segment + "/";
+            }
+
+            // First we add eclipse-target because the compiled classes are normally more updated
+            eclipseTargetLocations.add(new URL(eclipseApiLocation));
+            eclipseTargetLocations.add(new URL(apiLocation));
+
+            Model apiModel = parseModel(new File(apiLocationAsFile));
+
+            // Now we add the dependencies of API. We need this because if the API jar is not found on the
+            // local maven repo, this is the way to get its dependencies
+            eclipseTargetLocations.addAll(getDependenciesURLS(apiModel, eclipseTargetLocations, 1));
+        }
+
+        return eclipseTargetLocations;
+    }
+
+    /**
+     * Gets the byte code location, that is located on classes/ folder
+     * @param source
+     *            where to get the byte code location from
+     * @return File where the byte code is located
+     */
+    private File getByteCodeLocation(JavaSourceUsingMaven source) {
+        // Create a File object on the root directory of the classes
+        File byteCodeLocation = null;
+        String byteCodeLocationString = source.getByteCodeLocation().toString();
+        if (byteCodeLocationString.substring(byteCodeLocationString.lastIndexOf(File.separator) + 1)
+            .equals("test-classes")) {
+            byteCodeLocation =
+                source.getByteCodeLocation().getParentFile().toPath().resolve("classes" + File.separator).toFile();
+        } else {
+            byteCodeLocation = source.getByteCodeLocation();
+        }
+        return byteCodeLocation;
+    }
+
+    /**
+     * Retrieves dependencies from a Maven project. Up to 4 levels of recursiveness (gets the dependency of
+     * the dependency of the dependency...)
+     * @param model
+     *            parsed from the POM of a project
+     * @return POM dependencies in URL format
+     */
+    private ArrayList<URL> getDependenciesURLS(Model model) {
+        ArrayList<URL> dependenciesURLS = new ArrayList<>();
+        int recursiveness = 0; // We only want to recursively get 4 levels of dependency
+
+        dependenciesURLS = getDependenciesURLS(model, dependenciesURLS, recursiveness);
+        return dependenciesURLS;
+    }
+
+    /**
+     * Recursively gets dependencies from the model, and stores them on a list.
+     * @param model
+     *            parsed from the POM of a project
+     * @param previousURLs
+     *            list of previous dependencies URLs
+     * @param recursiveness
+     *            level of recursiveness. How deep we get into dependencies.
+     * @return POM dependencies in URL format
+     */
+    private ArrayList<URL> getDependenciesURLS(Model model, ArrayList<URL> previousURLs, int recursiveness) {
+
+        if (recursiveness >= 4) {
```

Comment:
I agree that it would be very nice to have a lazily loading of the dependencies. However, this seems a very complex topic. For instance:

When I try to load `EmployeeEntity`, I first need to load `Employee` defined on the `api`. This is easy because the `core` already defined `api` as a dependency. But, `Employee` extends `ApplicationEntity` which extends `GenericEntity`. That class is tricky to load because it is not defined inside the `api`.

The `api` defines as dependency `devon4j-rest` which defines as dependency `devon4j-basic`. So it is indirectly referenced.

If Maven was created, was in order to fix this [JAR hell](https://dzone.com/articles/what-is-jar-hell) problem.  If you run the command `mvn dependency:build-classpath` in any of your Maven projects, you will see all the jars that are loaded in the class path.

If you guide me on this, we can try to find an optimal solution. However, my proposed solution already works pretty well. We have even tested using My Thai Star, which as you know, is a large project.